lib/rex/post/meterpreter/ui/console/command_dispatcher/core.rb

```diff
@@ -1833,15 +1841,21 @@ def add_extension_client(mod)
     self.extensions << mod
   end
 
-  def tab_complete_postmods
-    tabs = client.framework.modules.post.map { |name,klass|
+  def tab_complete_modules
+    tabs = []
+    client.framework.modules.post.map do |name,klass|
       mod = client.framework.modules.post.create(name)
-      if mod and mod.session_compatible?(client)
-        mod.fullname.dup
-      else
-        nil
-      end
-    }
+      next unless mod && mod.session_compatible?(client)
+
+      tabs << mod.fullname.dup
+    end
+    client.framework.modules.exploits.each do |name,klass|
+      next unless klass && klass.ancestors.include?(Msf::Exploit::Local)
+      mod = client.framework.modules.exploits.create(name)
+      next unless mod && mod.session_compatible?(client)
```

Comment:
This seems to be sluggish and will fail to complete modules that currently show `[!] SESSION may not be compatible with this module.`, e.g `run exploit/osx/local/cfprefsd_race_condition`
I wonder if we could do something simpler like:
```
  def tab_complete_modules(str, words)
    tabs = []
    client.framework.modules.post.map do |name,klass|
      tabs << 'post/' + name
    end
    client.framework.modules.exploits.each do |name, klass|
      next unless klass && klass.ancestors.include?(Msf::Exploit::Local)
      tabs << 'exploit/' + name
    end
    return tabs.sort
  end
```

Reply:
So cut out the compatibility check? I'd be fine with that, I think I noticed that about the compatibility check as well. It's not particularly reliable.

With that change though, is the performance satisfactory?

Reply:
Oops actually no, it's the same, it's still really sluggish due to `client.framework.modules.exploits.each`
`client.framework.modules.module_names('exploit').each`
is 10x faster.

Reply:
Maybe we could do a hybrid approach? We could use
```ruby
client.framework.modules.module_names('exploit').grep(/\/(multi|#{Regexp.escape(client.platform)})\/local\//).each do |name|
  mod = client.framework.modules.exploits.create(name)
  next unless mod && mod.session_compatible?(client)
```
to reduce the number of modules we need to initialize and still produce accurate results. I'm not a huge fan of using module names to infer functionality, but baring a more useful module cache I think using them here for optimization if fine.

Reply:
Nice work @acammack-r7 , my only suggestion is we drop the session_compatible check.
How about something like: https://github.com/zeroSteiner/metasploit-framework/pull/13 ?

Reply:
That looks good if we want to drop the compatibility check. I like the idea of doing a check here (ex, for not showing meterpreter-only modules with shell sessions), but if it doesn't apply broadly then we don't need to keep it around.